tt_metal/impl/dispatch/command_queue.hpp

```diff
@@ -483,22 +502,27 @@ class HWCommandQueue {
     friend Trace BeginTrace(CommandQueue& cq);
     friend void EnqueueProgramImpl(CommandQueue& cq, std::variant < std::reference_wrapper<Program>, std::shared_ptr<Program> > program, bool blocking, std::optional<std::reference_wrapper<Trace>> trace);
     friend void EnqueueReadBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, void* dst, bool blocking);
-    friend void EnqueueWriteBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, const void* src, bool blocking);
+    friend void EnqueueWriteBufferImpl(CommandQueue& cq, std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer> > buffer, HostBufferMemTypes src, bool blocking);
+    friend void EnqueueAllocateBufferImpl(AllocBufferMetadata alloc_md);
+    friend void EnqueueDeallocateBufferImpl(AllocBufferMetadata alloc_md);
+    friend void EnqueueGetBufferAddrImpl(void* dst_buf_addr, const Buffer* buffer);
     friend void EnqueueRecordEventImpl(CommandQueue& cq, std::reference_wrapper<Event> event);
     friend void EnqueueWaitForEventImpl(CommandQueue& cq, std::reference_wrapper<Event> event);
     friend void FinishImpl(CommandQueue & cq);
     friend void EnqueueRecordEvent(CommandQueue& cq, Event &event);
     friend class Trace;
 };
 
-
 // Common interface for all command queue types
 struct CommandInterface {
     EnqueueCommandType type;
     std::optional<bool> blocking;
     std::optional<std::variant<std::reference_wrapper<Buffer>, std::shared_ptr<Buffer>>> buffer;
     std::optional<std::variant<std::reference_wrapper<Program>, std::shared_ptr<Program>>> program;
-    std::optional<const void*> src;
+    std::optional<AllocBufferMetadata> alloc_md;
+    std::optional<RuntimeArgsMetadata> runtime_args_md;
+    std::optional<const Buffer*> shadow_buffer;
+    std::optional<HostBufferMemTypes> src;
     std::optional<void*> dst;
```

Comment:
Not sure if this is needed here. The only reason we pass in `HostBufferMemTypes ` is because it contains `shared_ptr` types which allow the child thread to have access to data even when the parent gives up ownership.

When we read data into `dst`, it should remain allocated until the operation is complete. I can add this for cosmetic reasons, if you think it looks better.